<a href="https://colab.research.google.com/github/datascience-uniandes/data-quality-tutorial/blob/master/data-quality-tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Quality tutorial

MINE-4101: Applied Data Science  
Univerisdad de los Andes  
  
Dataset: Homicides Colombia  
Source: [datos.gov.co](datos.gov.co)
  
Last update: September, 2022

In [ ]:
!pip install pylev

In [1]:
# Importing required libraries

import re
from random import randint
from datetime import datetime

import numpy as np
import pandas as pd

import pylev

In [2]:
# Parameter for showing all columns when printing a dataframe
pd.set_option('display.max_columns', None)

In [3]:
# Loading data
homicides_df = pd.read_csv("./data/homicides.csv")

In [4]:
# Printing the dataset dimensions
homicides_df.shape

(12400, 22)

In [5]:
# Printing column data types
homicides_df.dtypes

FECHA                 object
DEPARTAMENTO          object
MUNICIPIO             object
DIA                   object
HORA                  object
BARRIO                object
ZONA                  object
CLASE DE SITIO        object
ARMA O MEDIO          object
MOVIL VICTIMA         object
MOVIL AGRESOR         object
EDAD                 float64
GENERO                object
ESTADO CIVIL          object
CLASE EMPLEADO        object
PROFESION             object
ESCOLARIDAD           object
PAIS NACE             object
CODIGO DANE          float64
AÑO DE NACIMIENTO    float64
CÉDULA                object
CORREO                object
dtype: object

In [6]:
homicides_df.head()

FECHA DEPARTAMENTO      MUNICIPIO     DIA   HORA  \
0  01/01/2021 12:00:00 AM    ANTIOQUIA          AMAGÁ  Jueves   6:00   
1  01/01/2021 12:00:00 AM    ANTIOQUIA        BARBOSA  Jueves   9:00   
2  01/01/2021 12:00:00 AM    ANTIOQUIA       EL BAGRE  Jueves  19:00   
3  01/01/2021 12:00:00 AM    ANTIOQUIA         JARDÍN  Jueves  11:20   
4  01/01/2021 12:00:00 AM    ANTIOQUIA  MEDELLÍN (CT)   Juees  15:00   

                     BARRIO    ZONA                      CLASE DE SITIO  \
0                 EL VOLCAN   RURAL                              TIENDA   
1            VDA. MATASANOS   RURAL                       VIAS PUBLICAS   
2             PUERTO CLAVER   RURAL                  FINCAS Y SIMILARES   
3                CRISTIANIA   RURAL                  FINCAS Y SIMILARES   
4  PICACHITO CNO REPORTADO6  URBANA  FRENTE A RESIDENCIAS - VIA PUBLICA   

   ARMA O MEDIO MOVIL VICTIMA MOVIL AGRESOR  EDAD     GENERO ESTADO CIVIL  \
0   ARMA BLANCA         A PIE         A PIE  44.0  MASCULINO       CASADO   
1   ARMA BLANCA         A PIE         A PIE  30.0  MASCULINO      SOLTERO   
2   ARMA BLANCA         A PIE         A PIE  33.0  MASCULINO  UNION LIBRE   
3   ARMA BLANCA         A PIE         A PIE  40.0  MASCULINO       CASADO   
4  CONTUNDENTES         A PIE         A PIE  66.0  MASCULINO  UNION LIBRE   

  CLASE EMPLEADO     PROFESION ESCOLARIDAD     PAIS NACE  CODIGO DANE  \
0  INDEPENDIENTE  NO REPORTADO    PRIMARIA      COLOMBIA    5030000.0   
1    DESEMPLEADO  NO REPORTADO  SECUNDARIA  NO REPORTADO    5079000.0   
2     AGRICULTOR  NO REPORTADO    PRIMARIA      COLOMBIA    5250000.0   
3     AGRICULTOR  NO REPORTADO    PRIMARIA      COLOMBIA    5364000.0   
4    DESEMPLEADO  NO REPORTADO    PRIMARIA      COLOMBIA    5001000.0   

   AÑO DE NACIMIENTO  CÉDULA                      CORREO  
0             1971.0  42-908  agbnqg2122@unidatos.edu.co  
1             1985.0  15-183        rbkeui3584@gmail.com  
2             1982.0  84-786  aorkhf9155@unidatos.edu.co  
3             1975.0  31-289  dhtemr6623@unidatos.edu.co  
4             1949.0  66-363  artatj9268@unidatos.edu.co

In [7]:
# Creating a lambda expression for datetime parsing
dateparse = lambda x: datetime.strptime(x, "%m/%d/%Y %H:%M:%S %p")

In [ ]:
# Applying the validation to all values in column
homicides_df["FECHA"].apply(dateparse)

# IT IS EXPECTED TO HAVE AN ERROR INTENTIONALLY

In [8]:
# Creating a function for validating which value is causing the previous error
def error_in_format(x):
    try:
        datetime.strptime(x, "%m/%d/%Y %H:%M:%S %p")
        return False
    except:
        return True

In [9]:
# Using the function for validation
homicides_df.loc[homicides_df["FECHA"].apply(error_in_format)]

# THE LAST ROW IS CAUSING THE ERROR

FECHA DEPARTAMENTO MUNICIPIO  DIA HORA BARRIO ZONA CLASE DE SITIO  \
12399  TOTAL          NaN       NaN  NaN  NaN    NaN  NaN            NaN   

      ARMA O MEDIO MOVIL VICTIMA MOVIL AGRESOR  EDAD GENERO ESTADO CIVIL  \
12399          NaN           NaN           NaN   NaN    NaN          NaN   

      CLASE EMPLEADO PROFESION ESCOLARIDAD PAIS NACE  CODIGO DANE  \
12399            NaN       NaN         NaN       NaN          NaN   

       AÑO DE NACIMIENTO CÉDULA CORREO  
12399                NaN    NaN    NaN

In [10]:
# Deleting a row by its index
homicides_df.drop(12399, inplace = True)

In [11]:
# Trying to parse the datetime string again
homicides_df["FECHA"] = homicides_df["FECHA"].apply(dateparse)

In [12]:
# Counting homicides by hour
homicides_df["FECHA"].dt.hour.value_counts()

# All datetime hour parts are the same

12    12399
Name: FECHA, dtype: int64

In [13]:
# The homicide hour is available in a different column!
# Merging both columns
homicides_df["FECHA"] = homicides_df["FECHA"].astype(str).apply(lambda x: x[:11]) + homicides_df["HORA"]

In [14]:
del homicides_df["HORA"]

In [15]:
homicides_df["FECHA"].head()

0     2021-01-01 6:00
1     2021-01-01 9:00
2    2021-01-01 19:00
3    2021-01-01 11:20
4    2021-01-01 15:00
Name: FECHA, dtype: object

In [16]:
# Making a new expression for datetime parsing
dateparse = lambda x: datetime.strptime(x, "%Y-%m-%d %H:%M")

In [17]:
# Applying the expression
homicides_df["FECHA"] = homicides_df["FECHA"].apply(dateparse)

In [18]:
homicides_df.dtypes

FECHA                datetime64[ns]
DEPARTAMENTO                 object
MUNICIPIO                    object
DIA                          object
BARRIO                       object
ZONA                         object
CLASE DE SITIO               object
ARMA O MEDIO                 object
MOVIL VICTIMA                object
MOVIL AGRESOR                object
EDAD                        float64
GENERO                       object
ESTADO CIVIL                 object
CLASE EMPLEADO               object
PROFESION                    object
ESCOLARIDAD                  object
PAIS NACE                    object
CODIGO DANE                 float64
AÑO DE NACIMIENTO           float64
CÉDULA                       object
CORREO                       object
dtype: object

In [19]:
homicides_df.head()

FECHA DEPARTAMENTO      MUNICIPIO     DIA  \
0 2021-01-01 06:00:00    ANTIOQUIA          AMAGÁ  Jueves   
1 2021-01-01 09:00:00    ANTIOQUIA        BARBOSA  Jueves   
2 2021-01-01 19:00:00    ANTIOQUIA       EL BAGRE  Jueves   
3 2021-01-01 11:20:00    ANTIOQUIA         JARDÍN  Jueves   
4 2021-01-01 15:00:00    ANTIOQUIA  MEDELLÍN (CT)   Juees   

                     BARRIO    ZONA                      CLASE DE SITIO  \
0                 EL VOLCAN   RURAL                              TIENDA   
1            VDA. MATASANOS   RURAL                       VIAS PUBLICAS   
2             PUERTO CLAVER   RURAL                  FINCAS Y SIMILARES   
3                CRISTIANIA   RURAL                  FINCAS Y SIMILARES   
4  PICACHITO CNO REPORTADO6  URBANA  FRENTE A RESIDENCIAS - VIA PUBLICA   

   ARMA O MEDIO MOVIL VICTIMA MOVIL AGRESOR  EDAD     GENERO ESTADO CIVIL  \
0   ARMA BLANCA         A PIE         A PIE  44.0  MASCULINO       CASADO   
1   ARMA BLANCA         A PIE         A PIE  30.0  MASCULINO      SOLTERO   
2   ARMA BLANCA         A PIE         A PIE  33.0  MASCULINO  UNION LIBRE   
3   ARMA BLANCA         A PIE         A PIE  40.0  MASCULINO       CASADO   
4  CONTUNDENTES         A PIE         A PIE  66.0  MASCULINO  UNION LIBRE   

  CLASE EMPLEADO     PROFESION ESCOLARIDAD     PAIS NACE  CODIGO DANE  \
0  INDEPENDIENTE  NO REPORTADO    PRIMARIA      COLOMBIA    5030000.0   
1    DESEMPLEADO  NO REPORTADO  SECUNDARIA  NO REPORTADO    5079000.0   
2     AGRICULTOR  NO REPORTADO    PRIMARIA      COLOMBIA    5250000.0   
3     AGRICULTOR  NO REPORTADO    PRIMARIA      COLOMBIA    5364000.0   
4    DESEMPLEADO  NO REPORTADO    PRIMARIA      COLOMBIA    5001000.0   

   AÑO DE NACIMIENTO  CÉDULA                      CORREO  
0             1971.0  42-908  agbnqg2122@unidatos.edu.co  
1             1985.0  15-183        rbkeui3584@gmail.com  
2             1982.0  84-786  aorkhf9155@unidatos.edu.co  
3             1975.0  31-289  dhtemr6623@unidatos.edu.co  
4             1949.0  66-363  artatj9268@unidatos.edu.co

In [20]:
# Creating a dictionary representing the valid departments for Colombia
departments_list = ['ANTIOQUIA', 'ATLÁNTICO', 'BOLÍVAR', 'BOYACÁ', 'CALDAS', 'CAQUETÁ',
       'CASANARE', 'CAUCA', 'CESAR', 'CHOCÓ', 'CÓRDOBA', 'META',
       'CUNDINAMARCA', 'HUILA', 'MAGDALENA', 'NARIÑO', 'PUTUMAYO',
       'RISARALDA', 'SANTANDER', 'SUCRE', 'TOLIMA', 'VALLE',
       'NORTE DE SANTANDER', 'GUAJIRA', 'QUINDÍO', 'SAN ANDRÉS Y PROVIDENCIA', 'ARAUCA',
       'GUAINÍA', 'VICHADA', 'VAUPÉS', 'GUAVIARE', 'AMAZONAS']

In [21]:
# Finding values not matching with the dictionary
homicides_df.loc[~homicides_df["DEPARTAMENTO"].isin(departments_list), "DEPARTAMENTO"].unique()

array(['SAN ANDRÉS', 'N. DE SANTANDER'], dtype=object)

<span style="color:red">TODO: Replace the values identified as error to a valid value from the dictionary</span>

<span style="color:red">Hint: You can use the replace() pandas function</span>

In [22]:
homicides_df.DEPARTAMENTO = homicides_df.DEPARTAMENTO.replace({'SAN ANDRÉS':'SAN ANDRÉS Y PROVIDENCIA','N. DE SANTANDER':'NORTE DE SANTANDER'})

In [23]:
homicides_df.loc[~homicides_df["DEPARTAMENTO"].isin(departments_list), "DEPARTAMENTO"].unique()

array([], dtype=object)

In [24]:
homicides_df.DEPARTAMENTO.unique()

array(['ANTIOQUIA', 'ATLÁNTICO', 'BOLÍVAR', 'BOYACÁ', 'CALDAS', 'CAQUETÁ',
       'CASANARE', 'CAUCA', 'CESAR', 'CHOCÓ', 'CÓRDOBA', 'META',
       'CUNDINAMARCA', 'HUILA', 'MAGDALENA', 'NARIÑO', 'PUTUMAYO',
       'RISARALDA', 'SANTANDER', 'SUCRE', 'TOLIMA', 'VALLE',
       'NORTE DE SANTANDER', 'GUAJIRA', 'QUINDÍO',
       'SAN ANDRÉS Y PROVIDENCIA', 'ARAUCA', 'GUAINÍA', 'VICHADA',
       'GUAVIARE', 'VAUPÉS', 'AMAZONAS'], dtype=object)

In [25]:
# Detecting duplicates by "CÉDULA" column
duplicates_by_cedula = homicides_df.loc[homicides_df["CÉDULA"].duplicated(keep = False)]

In [26]:
duplicates_by_cedula.shape

(1627, 21)

In [27]:
# Showing some examples
duplicates_by_cedula.sort_values("CÉDULA", ascending = True).head(6)

FECHA DEPARTAMENTO      MUNICIPIO      DIA  \
4041  2021-05-03 23:30:00        CAUCA       EL TAMBO  Domingo   
3810  2021-04-26 03:00:00    ANTIOQUIA   SAN JERÓNIMO  Domingo   
693   2021-01-19 05:30:00    ANTIOQUIA         SALGAR    Lunes   
10796 2021-11-20 03:00:00        VALLE      CALI (CT)  Viernes   
2386  2021-03-14 02:00:00         META  FUENTE DE ORO   Sábado   
138   2021-01-02 06:20:00        CAUCA         SOTARA  Viernes   

                    BARRIO    ZONA                      CLASE DE SITIO  \
4041           LA VICTORIA   RURAL         BARES, CANTINAS Y SIMILARES   
3810              LA PLAYA  URBANA  HOTELES, RESIDENCIAS, Y SIMILARES.   
693              LA HABANA  URBANA                       VIAS PUBLICAS   
10796   POTRERO GRANDE E21  URBANA                       VIAS PUBLICAS   
2386   VEREDA PUERTO NUEVO   RURAL         BARES, CANTINAS Y SIMILARES   
138                 CENTRO   RURAL                       VIAS PUBLICAS   

             ARMA O MEDIO MOVIL VICTIMA MOVIL AGRESOR  EDAD     GENERO  \
4041        ARMA DE FUEGO         A PIE         A PIE  45.0  MASCULINO   
3810   CUERDA/SOGA/CADENA         A PIE         A PIE  25.0  MASCULINO   
693         ARMA DE FUEGO         A PIE         A PIE  58.0  MASCULINO   
10796       ARMA DE FUEGO         A PIE         A PIE  21.0  MASCULINO   
2386          ARMA BLANCA         A PIE         A PIE  42.0  MASCULINO   
138           ARMA BLANCA         A PIE         A PIE  22.0  MASCULINO   

      ESTADO CIVIL       CLASE EMPLEADO     PROFESION ESCOLARIDAD PAIS NACE  \
4041   UNION LIBRE           AGRICULTOR  NO REPORTADO    PRIMARIA  COLOMBIA   
3810       SOLTERO        INDEPENDIENTE  NO REPORTADO  SECUNDARIA  COLOMBIA   
693        SOLTERO           AGRICULTOR  NO REPORTADO    PRIMARIA  COLOMBIA   
10796      SOLTERO  EMPLEADO PARTICULAR  NO REPORTADO  SECUNDARIA  COLOMBIA   
2386       SOLTERO        INDEPENDIENTE  NO REPORTADO    PRIMARIA  COLOMBIA   
138        SOLTERO    EMPLEADO EJERCITO  NO REPORTADO  SECUNDARIA  COLOMBIA   

       CODIGO DANE  AÑO DE NACIMIENTO  CÉDULA                      CORREO  
4041    19256000.0             1970.0  10-048        aitufn1227@gmail.com  
3810     5656000.0             1990.0  10-048  ibbcpu2509@unidatos.edu.co  
693      5642000.0               57.0  10-079  unhoqj1172@unidatos.edu.co  
10796   76001000.0             1994.0  10-079  iknnoj8430@unidatos.edu.co  
2386    50287000.0             1973.0  10-255  afompq7113@unidatos.edu.co  
138     19760000.0             1993.0  10-255  cobgqs8819@unidatos.edu.co

<span style="color:red">TODO: Delete records with "CÉDULA" duplicated</span>

<span style="color:red">Hint: You can use drop_duplicates() pandas function.</span>

In [28]:
homicides_df = homicides_df.drop_duplicates(subset = "CÉDULA")

In [29]:
duplicates_by_cedula = homicides_df.loc[homicides_df["CÉDULA"].duplicated(keep = False)]

Se eliminan los duplicados exitosamente

In [30]:
duplicates_by_cedula.shape

(0, 21)

In [31]:
# Using regular expressions for validating if "CÉDULA" values match the pattern XX-XXX
cedula_malformed = homicides_df.loc[homicides_df["CÉDULA"].apply(lambda x: (re.match("\d{2}-\d{3}", x) is None))]

In [32]:
cedula_malformed.shape

(28, 21)

In [33]:
cedula_malformed.head(6)

FECHA DEPARTAMENTO          MUNICIPIO      DIA  \
172  2021-01-03 11:00:00         META            MESETAS   Sábado   
1114 2021-02-01 12:30:00    ATLÁNTICO  BARRANQUILLA (CT)  Domingo   
2119 2021-03-06 15:30:00        VALLE          CALI (CT)  Viernes   
3309 2021-04-11 10:30:00      CAQUETÁ     FLORENCIA (CT)   Sávado   
3409 2021-04-13 23:20:00      CÓRDOBA            SAHAGÚN    Lunes   
3588 2021-04-19 20:00:00      BOLÍVAR     CARTAGENA (CT)  Domingo   

                            BARRIO    ZONA         CLASE DE SITIO  \
172                VEREDA EL CAFRE   RURAL         ZONA SELVÁTICA   
1114                        LA LUZ  URBANA          VIAS PUBLICAS   
2119           QUINTAS DEL SOL E14  URBANA  DENTRO DE LA VIVIENDA   
3309                   VIA MORELIA   RURAL               CARCELES   
3409  CORREGIMIENTO DE BAJO GRANDE  URBANA               BILLARES   
3588                  LA ESPERANZA  URBANA          VIAS PUBLICAS   

          ARMA O MEDIO MOVIL VICTIMA MOVIL AGRESOR  EDAD     GENERO  \
172   MINA ANTIPERSONA         A PIE         A PIE  24.0  MASCULINO   
1114     ARMA DE FUEGO         A PIE         A PIE  29.0  MASCULINO   
2119       ARMA BLANCA         A PIE         A PIE  20.0   FEMENINO   
3309         CORTANTES         A PIE         A PIE  23.0  MASCULINO   
3409      CONTUNDENTES         A PIE         A PIE  49.0  MASCULINO   
3588     ARMA DE FUEGO         A PIE         A PIE  45.0  MASCULINO   

     ESTADO CIVIL       CLASE EMPLEADO     PROFESION   ESCOLARIDAD  \
172       SOLTERO    EMPLEADO EJERCITO  NO REPORTADO    SECUNDARIA   
1114      SOLTERO        INDEPENDIENTE  NO REPORTADO      PRIMARIA   
2119  UNION LIBRE  EMPLEADO PARTICULAR  NO REPORTADO    SECUNDARIA   
3309      SOLTERO        INDEPENDIENTE  NO REPORTADO      PRIMARIA   
3409      SOLTERO        INDEPENDIENTE  NO REPORTADO    SECUNDARIA   
3588      SOLTERO          DESEMPLEADO  NO REPORTADO  NO REPORTADO   

         PAIS NACE  CODIGO DANE  AÑO DE NACIMIENTO  CÉDULA  \
172       COLOMBIA   50330000.0             1991.0  680-21   
1114      COLOMBIA    8001000.0             1986.0  140-17   
2119      COLOMBIA   76001000.0             1995.0  975-31   
3309      COLOMBIA   18001000.0             1992.0  348-66   
3409      COLOMBIA   23660000.0             1966.0  496-18   
3588  NO REPORTADO   13001000.0             1970.0  188-03   

                          CORREO  
172         oohghd8899@gmail.com  
1114  correo5853@colombia.gov.co  
2119  ohbqrk3631@unidatos.edu.co  
3309  pdkqur8407@unidatos.edu.co  
3409  fdbbeo6751@unidatos.edu.co  
3588  diebuo5651@unidatos.edu.co

<span style="color:red">TODO: Fix the malformed "CÉDULA" values.</span>

In [34]:
def cedula_format(x):
    arr = "".join(x.split("-"))
    return arr[0:2] + "-" + arr[2:]

In [35]:
cedula_malformed.CÉDULA = cedula_malformed.CÉDULA.apply(cedula_format)

/var/folders/v6/hwjhxff57kz0r3_zmmhctqk00000gn/T/ipykernel_1432/3745306228.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cedula_malformed.CÉDULA = cedula_malformed.CÉDULA.apply(cedula_format)


In [36]:
homicides_df.loc[cedula_malformed.index] = cedula_malformed

Se arregla el formato de todas las cedulas mal formadas

In [37]:
homicides_df.loc[homicides_df["CÉDULA"].apply(lambda x: (re.match("\d{2}-\d{3}", x) is None))].shape

(0, 21)

<span style="color:red">TODO: Make something similar to check and fix the "CORREO" column (PATTERN: 4 digits before the @, only .edu.co and .com domains are allowed)</span>

Hacemos una expresión regular para encontrar estos correos de la siguiente forma:
<ul>
<li>[a-zA-Z]* para que 0 o más letras al inicio</li>
<li>\d{4} seguido de 4 digitos como indica el enunciado antes del @</li>
</ul>
Sabiendo que tiene que terminar en .edu.co creamos la siguiente expresión regular:

In [38]:
malformed_correos = homicides_df.loc[homicides_df["CORREO"].apply(lambda x: (re.match("[a-zA-Z]*\d{4}@[a-zA-Z]*.edu.co", x) is None))]

In [39]:
malformed_correos.shape

(6258, 21)

In [40]:
malformed_correos.head()

FECHA DEPARTAMENTO      MUNICIPIO     DIA  \
1  2021-01-01 09:00:00    ANTIOQUIA        BARBOSA  Jueves   
6  2021-01-01 11:00:00    ANTIOQUIA  MEDELLÍN (CT)  Jueves   
7  2021-01-01 19:35:00    ANTIOQUIA  MEDELLÍN (CT)  Jueves   
13 2021-01-01 14:00:00    ANTIOQUIA        TÁMESIS  Jueves   
19 2021-01-01 08:30:00    ATLÁNTICO        SOLEDAD   juves   

                        BARRIO    ZONA      CLASE DE SITIO   ARMA O MEDIO  \
1               VDA. MATASANOS   RURAL       VIAS PUBLICAS    ARMA BLANCA   
6   CRISTO REY CNO REPORTADO15  URBANA       VIAS PUBLICAS    ARMA BLANCA   
7     LA SALLE  CNO REPORTADO3  URBANA       VIAS PUBLICAS    ARMA BLANCA   
13                   SAN PABLO   RURAL  FINCAS Y SIMILARES    ARMA BLANCA   
19                   LAS MORAS  URBANA       VIAS PUBLICAS  ARMA DE FUEGO   

   MOVIL VICTIMA MOVIL AGRESOR  EDAD     GENERO ESTADO CIVIL CLASE EMPLEADO  \
1          A PIE         A PIE  30.0  MASCULINO      SOLTERO    DESEMPLEADO   
6          A PIE         A PIE  25.0  MASCULINO      SOLTERO    DESEMPLEADO   
7          A PIE         A PIE  37.0  MASCULINO       CASADO  INDEPENDIENTE   
13         A PIE         A PIE  40.0  MASCULINO      SOLTERO     AGRICULTOR   
19         A PIE         A PIE  33.0  MASCULINO      SOLTERO  INDEPENDIENTE   

       PROFESION ESCOLARIDAD     PAIS NACE  CODIGO DANE  AÑO DE NACIMIENTO  \
1   NO REPORTADO  SECUNDARIA  NO REPORTADO    5079000.0             1985.0   
6   NO REPORTADO  SECUNDARIA      COLOMBIA    5001000.0             1990.0   
7   NO REPORTADO  SECUNDARIA      COLOMBIA    5001000.0             1978.0   
13  NO REPORTADO    PRIMARIA      COLOMBIA    5789000.0             1975.0   
19  NO REPORTADO    PRIMARIA      COLOMBIA    8758000.0             1982.0   

    CÉDULA                CORREO  
1   15-183  rbkeui3584@gmail.com  
6   45-052  pkdakc6682@gmail.com  
7   45-298  qledtq1884@gmail.com  
13  39-330  nsirqk5181@gmail.com  
19  52-287  shfbpo7917@gmail.com

In [41]:
malformed_correos.tail()

FECHA DEPARTAMENTO   MUNICIPIO     DIA  \
12392 2021-12-31 11:30:00        VALLE   CALI (CT)  Jueves   
12394 2021-12-31 15:42:00        VALLE  EL CERRITO  Jueves   
12395 2021-12-31 19:04:00        VALLE   LA CUMBRE  Jueves   
12397 2021-12-31 14:30:00        VALLE      ZARZAL  Jueves   
12398 2021-12-31 22:55:00      VICHADA    CUMARIBO  Jueves   

                        BARRIO    ZONA CLASE DE SITIO   ARMA O MEDIO  \
12392  BRISAS DE LOS ALAMOS E2  URBANA  VIAS PUBLICAS  ARMA DE FUEGO   
12394             LA ESPERANZA  URBANA  VIAS PUBLICAS  ARMA DE FUEGO   
12395          CGTO BITACO E27   RURAL  VIAS PUBLICAS  ARMA DE FUEGO   
12397        URB. EL ZARZALITO  URBANA  VIAS PUBLICAS  ARMA DE FUEGO   
12398                EL CENTRO  URBANA     HOSPITALES    ARMA BLANCA   

      MOVIL VICTIMA MOVIL AGRESOR  EDAD     GENERO ESTADO CIVIL  \
12392         A PIE         A PIE  62.0  MASCULINO       CASADO   
12394         A PIE         A PIE  53.0   FEMENINO       CASADO   
12395         A PIE         A PIE  58.0  MASCULINO  UNION LIBRE   
12397         A PIE         A PIE  27.0  MASCULINO      SOLTERO   
12398         A PIE         A PIE  23.0  MASCULINO      SOLTERO   

            CLASE EMPLEADO     PROFESION ESCOLARIDAD PAIS NACE  CODIGO DANE  \
12392  EMPLEADO PARTICULAR  NO REPORTADO  SECUNDARIA  COLOMBIA   76001000.0   
12394  EMPLEADO PARTICULAR  NO REPORTADO    PRIMARIA  COLOMBIA   76248000.0   
12395           AGRICULTOR  NO REPORTADO    PRIMARIA  COLOMBIA   76377000.0   
12397  EMPLEADO PARTICULAR  NO REPORTADO  SECUNDARIA  COLOMBIA   76895000.0   
12398        INDEPENDIENTE  NO REPORTADO  SECUNDARIA  COLOMBIA   99773000.0   

       AÑO DE NACIMIENTO  CÉDULA                      CORREO  
12392             1953.0  91-774   correo295@unidatos.edu.co  
12394             1962.0  39-658        nggiso7664@gmail.com  
12395             1957.0  39-442  correo4352@colombia.gov.co  
12397             1988.0  62-735        gafkup6393@gmail.com  
12398             1992.0  47-126   correo736@unidatos.edu.co

Podemos ver que en el resultado se muestran correos que no terminan con el dominio .edu.co o que tienen menos de 4 numeros antes del @.

El enunciado no dice como "corregir" estos problemas. Una solución podria ser eliminar esos correos, no la fila entera porque se deberian eliminar demasiados datos ya que como vimos son 6258 filas. Sin embargo <b> esto depende del negocio </b> por lo que se decide solo visualizarlos y no realizar ningún procedimiento adicional

In [42]:
# Showing different values for "DIA" column
homicides_df["DIA"].unique()

array(['Jueves', 'Juees', 'Jueces', 'juves', 'Juevrs', 'Viernes',
       'Viermes', 'iernes', 'virnes', 'Vierens', 'Sábado', 'Sabadi',
       'Sabado', 'sábad', 'Sávado', 'Ssbado', 'Domingo', 'Domungo',
       'Doningo', 'domungo', 'Lunes', 'lune', 'Luns', 'Lumes', 'kunes',
       'Lnues', 'Martes', 'Mates', 'Marte', 'mates', 'Miércoles',
       'Miwrcoles', 'Mircoles', 'Voernes', 'domnigo', 'Maryes',
       'Miercoles', 'miércles', 'Dominog', 'Msrtes', 'Mirrcoles'],
      dtype=object)

In [43]:
# Calculating the distance between two words using the Levenshtein method
pylev.levenshtein('sábado', 'sabaod')

3

In [44]:
pylev.levenshtein('sábado', 'viernes')

7

<span style="color:red">TODO: Create a function to fix the digitation errors for "DIA" column.</span>

In [45]:
dias = ["Lunes","Martes","Miércoles","Jueves","Viernes","Sábado","Domingo"]

def check_day(x):
    leven = 100
    resp = ""
    for dia in dias:
        distancia = pylev.levenshtein(dia, x)
        if distancia < leven:
            leven = distancia
            resp = dia
    return resp

In [46]:
homicides_df["DIA"] = homicides_df["DIA"].apply(check_day)

In [47]:
homicides_df["DIA"].unique()

array(['Jueves', 'Lunes', 'Viernes', 'Sábado', 'Domingo', 'Martes',
       'Miércoles'], dtype=object)

In [48]:
homicides_df[["AÑO DE NACIMIENTO", "EDAD"]].head(10)

AÑO DE NACIMIENTO  EDAD
0             1971.0  44.0
1             1985.0  30.0
2             1982.0  33.0
3             1975.0  40.0
4             1949.0  66.0
5             1973.0  42.0
6             1990.0  25.0
7             1978.0  37.0
8             1990.0  25.0
9             1985.0  30.0

<span style="color:red">TODO: Fix the "AÑO DE NACIMIENTO" column using the column "EDAD".</span>

In [49]:
def anio_correcto(x):
    #Asumiendo que los datos fueron tomados en 2022
    return 2022 - x

In [50]:
homicides_df["AÑO DE NACIMIENTO"] = homicides_df.EDAD.apply(anio_correcto)

In [51]:
homicides_df[["AÑO DE NACIMIENTO", "EDAD"]].head(10)

AÑO DE NACIMIENTO  EDAD
0             1978.0  44.0
1             1992.0  30.0
2             1989.0  33.0
3             1982.0  40.0
4             1956.0  66.0
5             1980.0  42.0
6             1997.0  25.0
7             1985.0  37.0
8             1997.0  25.0
9             1992.0  30.0